osu.Game.Rulesets.Osu/Difficulty/Skills/OsuSkill.cs

```diff
@@ -0,0 +1,411 @@
+            };
+
+            for (int i = first; i <= last; ++i)
+            {
+                var note = noteDifficulties[i];
+                double hitProbability = fcProbabilityPrecomputed(powSkill, note.ExponentiatedDifficulty) + 1e-10;
+                result.ExpectedTimeUntilFullCombo = (result.ExpectedTimeUntilFullCombo + note.DeltaTime) / hitProbability;
+            }
+
+            return result;
+        }
+
+        /// <summary>
+        /// The estimated time it takes a player to full-combo a particular submap,
+        /// assuming they retry as soon as they miss a note.
+        /// </summary>
+        private MapSectionPerformanceData aggregateMapSectionPerformanceData(MapSectionPerformanceData[] sectionData, int first, int count)
+        {
+            int last = first + count - 1;
+
+            // TODO: clarify - is this just the approximate time needed to restart the map?
+            var result = new MapSectionPerformanceData
+            {
+                ExpectedTimeUntilFullCombo = map_retry_time,
+                FullComboProbability = 1,
+                StartTime = sectionData[first].StartTime,
+                EndTime = sectionData[last].EndTime
+            };
+
+            for (int i = first; i <= last; ++i)
+            {
+                // simplification - assumes that two sections being hit are independent events
+                // (not necessarily the case, but that is hard to estimate objectively)
+                result.ExpectedTimeUntilFullCombo /= sectionData[i].FullComboProbability;
+                result.ExpectedTimeUntilFullCombo += sectionData[i].ExpectedTimeUntilFullCombo;
+                result.FullComboProbability *= sectionData[i].FullComboProbability;
+                result.ExponentiatedDifficulty += sectionData[i].ExponentiatedDifficulty;
+            }
+
+            return result;
+        }
+
+        /// <summary>
+        /// Estimates the number of misses expected from players that can FC maps <see cref="MISS_STAR_RATING_MULTIPLIERS"/> as hard as the current one.
+        /// The returned value corresponds directly to <see cref="MissCounts"/>.
+        /// </summary>
+        /// <param name="fullComboDifficulty">The star difficulty of a full combo on the current map.</param>
+        private double[] calculateMissCounts(double fullComboDifficulty)
+        {
+            var result = new double[difficulty_count];
+
+            double fcSkill = fullComboDifficulty / fullComboDifficultyMultiplier;
+            double fcProb = fcProbability(fcSkill, fullComboDifficulty);
+
+            for (int i = 0; i < difficulty_count; ++i)
+            {
+                double missDifficulty = fullComboDifficulty * MISS_STAR_RATING_MULTIPLIERS[i];
+                double missSkill = missDifficulty / fullComboDifficultyMultiplier;
+
+                double[] missProbabilities = getMissProbabilities(missSkill);
+                result[i] = getMissCount(fcProb, missProbabilities);
```

Comment:
This is one thing that seemed slightly weird in my pass - statistics was never my good side, but I'm not seeing the connection between the full combo probability and the per-note miss probabilities. The methodology (finding the zero of the CDF) looks correct, but I'm just not sure why this is the threshold chosen.

@joseph-ireland can you advise on this if you ever get a moment?

Reply:
The full combo probability comes from the previous step - finding the skill that can FC with the given amount of time spent on failed attempts.

Using that as a threshold will make the 100% combo and the 0 miss values match up.

You could also interpret it as including the length of the map in there. 

successProbability = expectedTimePerPlay/expectedTimeUntilSuccess, so keeping it fixed is essentially assuming that the time spent per play of the map is the same for different skill levels, and they just allow themselves to make more misses.

Reply:
Right, that does make sense, thanks!

Reply:
I mentioned this to bdach in a PM, but I am still available for consulting in this thread. Taking a break primarily from the community relations aspect.